### PR TITLE
Fixed #32984 -- Allowed customizing a deletion field widget in formsets.

### PR DESCRIPTION
**Pull request by tiesjan**
_ticket-32984_

Closely modeled after the custom ordering widget implementation.

### CODE REVIEW

File: tests/forms_tests/tests/test_formsets.py [hunk 1]
```diff
@@ -602,16 +630,19 @@ def test_formsets_with_ordering(self):
             ],
         )
 
-    def test_formsets_with_order_custom_widget(self):
-        class OrderingAttributFormSet(BaseFormSet):
+    def test_formsets_with_ordering_custom_widget(self):
+        """
+        Asserts that the custom ordering widget is correctly applied.
+        """
```
Comment (felixxm) [inline tests/forms_tests/tests/test_formsets.py:636]: To add more explanation, docstrings should state the expected behavior and omit prefixes like "Asserts that" since all tests assert things. This guideline is from [Python coding style](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/#python-style). TBH, I don't see much value in this docstring, please revert it.

File: tests/forms_tests/tests/test_formsets.py [hunk 2]
```diff
@@ -602,16 +630,19 @@ def test_formsets_with_ordering(self):
...
+        class OrderingAttributeFormSet(BaseFormSet):
```
Comment (felixxm) [inline tests/forms_tests/tests/test_formsets.py:637]: Please move fixing typo to a separate commit.

File: django/forms/formsets.py [hunk 3]
```diff
@@ -417,7 +422,11 @@ def add_fields(self, form, index):
                     widget=self.get_ordering_widget(),
                 )
         if self.can_delete and (self.can_delete_extra or index < initial_form_count):
-            form.fields[DELETION_FIELD_NAME] = BooleanField(label=_('Delete'), required=False)
+            form.fields[DELETION_FIELD_NAME] = BooleanField(
+                label=_('Delete'),
+                required=False,
+                widget=self.get_deletion_widget()
```
Comment (felixxm) [inline django/forms/formsets.py:428]:
```suggestion
                widget=self.get_deletion_widget(),
```

File: docs/releases/4.0.txt [hunk 4]
```diff
@@ -229,6 +229,12 @@ Forms
   an additional class of ``nonform`` to help distinguish them from
   form-specific errors.
 
+* :class:`~django.forms.formsets.BaseFormSet` now allows controlling the widget
```
Comment (felixxm) [inline docs/releases/4.0.txt:232]:
```suggestion
* :class:`~django.forms.formsets.BaseFormSet` now allows customizing the widget
```

File: docs/releases/4.0.txt [hunk 5]
```diff
@@ -229,6 +229,12 @@ Forms
...
+  used when deleting forms via
+  :attr:`~django.forms.formsets.BaseFormSet.can_delete` by setting the
+  :attr:`~django.forms.formsets.BaseFormSet.deletion_widget` attribute or
+  overriding :attr:`~django.forms.formsets.BaseFormSet.get_deletion_widget()`.
```
Comment (felixxm) [inline docs/releases/4.0.txt:236]:
```suggestion
  overriding :meth:`~django.forms.formsets.BaseFormSet.get_deletion_widget` method.
```

File: tests/forms_tests/tests/test_formsets.py [hunk 6]
```diff
@@ -551,6 +551,34 @@ def test_formset_with_deletion_invalid_deleted_form(self):
...
+        for formset_class, delete_html in tests:
+            with self.subTest(formset_class=formset_class.__name__):
+                ArticleFormSet = formset_factory(ArticleForm, formset=formset_class, can_delete=True)
+                formset = ArticleFormSet(auto_id=False)
+                self.assertHTMLEqual(
+                    '\n'.join(form.as_ul() for form in formset.forms),
+                    (
+                        '<li>Title: <input type="text" name="form-0-title"></li>'
+                        '<li>Pub date: <input type="text" name="form-0-pub_date">'
+                        '%s</li>' % delete_html
```
Comment (felixxm) [inline tests/forms_tests/tests/test_formsets.py:578]:
```suggestion
                        f'<li>Title: <input type="text" name="form-0-title"></li>'
                        f'<li>Pub date: <input type="text" name="form-0-pub_date">'
                        f'{delete_form}</li>'
```

File: tests/forms_tests/tests/test_formsets.py [hunk 7]
```diff
@@ -551,6 +551,34 @@ def test_formset_with_deletion_invalid_deleted_form(self):
...
+        tests = (
+            (DeletionAttributeFormSet, '<input type="hidden" name="form-0-DELETE">'),
+            (DeletionMethodFormSet, '<input class="deletion" type="hidden" name="form-0-DELETE">'),
+        )
+        for formset_class, delete_html in tests:
+            with self.subTest(formset_class=formset_class.__name__):
+                ArticleFormSet = formset_factory(ArticleForm, formset=formset_class, can_delete=True)
+                formset = ArticleFormSet(auto_id=False)
+                self.assertHTMLEqual(
+                    '\n'.join(form.as_ul() for form in formset.forms),
```
Comment (felixxm) [inline tests/forms_tests/tests/test_formsets.py:574]:
```suggestion
                    '\n'.join([form.as_ul() for form in formset.forms]),
```

File: tests/forms_tests/tests/test_formsets.py [hunk 8]
```diff
@@ -551,6 +551,34 @@ def test_formset_with_deletion_invalid_deleted_form(self):
         self.assertEqual(formset._errors, [])
         self.assertEqual(len(formset.deleted_forms), 1)
 
+    def test_formset_with_deletion_custom_widget(self):
+        """
+        Asserts that the custom deletion widget is correctly applied.
+        """
```
Comment (felixxm) [inline tests/forms_tests/tests/test_formsets.py:557]: I'd chop this docstring.

**Review by felixxm — COMMENTED** [review]
@tiesjan Thanks for this patch :+1: I left small comments.

**Comment by tiesjan** [comment]
Thanks for the helpful feedback, @felixxm! I think I've addressed all of it.

**Comment by felixxm** [comment]
@tiesjan Thanks :+1:

I pushed minor edits.
